plugins/modem-manager/fu-sahara-loader.c

```diff
@@ -0,0 +1,562 @@
+
+	FuUsbDevice *usb_device;
+
+	int ep_in;
+	int ep_out;
+	gsize maxpktsize_in;
+	gsize maxpktsize_out;
+};
+
+G_DEFINE_TYPE(FuSaharaLoader, fu_sahara_loader, G_TYPE_OBJECT)
+
+/* Protocol definitions */
+typedef enum {
+	SAHARA_NO_CMD_ID = 0,
+	SAHARA_HELLO_ID,
+	SAHARA_HELLO_RESPONSE_ID,
+	SAHARA_READ_DATA_ID,
+	SAHARA_END_OF_IMAGE_TX_ID,
+	SAHARA_DONE_ID,
+	SAHARA_DONE_RESP_ID,
+	SAHARA_RESET_ID,
+	SAHARA_RESET_RESPONSE_ID,
+	SAHARA_READ_DATA_64_BIT_ID = 0x12,
+	SAHARA_LAST_CMD_ID
+} FuSaharaCommandId;
+
+typedef enum {
+	SAHARA_STATUS_SUCCESS = 0,
+	SAHARA_STATUS_FAILED,
+	SAHARA_STATUS_LAST
+} FuSaharaStatusCode;
+
+typedef enum {
+	SAHARA_MODE_IMAGE_TX_PENDING,
+	SAHARA_MODE_IMAGE_TX_COMPLETE,
+	SAHARA_MODE_LAST
+} FuSaharaMode;
+
+/* Sahara packet definition */
+struct sahara_packet {
+	guint32 command_id;
+	guint32 length;
+
+	union {
+		struct {
+			guint32 version;
+			guint32 version_compatible;
+			guint32 max_packet_length;
+			guint32 mode;
+		} hello;
+		struct {
+			guint32 version;
+			guint32 version_compatible;
+			guint32 status;
+			guint32 mode;
+			guint32 reserved[6];
+		} hello_response;
+		struct {
+			guint32 image_id;
+			guint32 offset;
+			guint32 length;
+		} read_data;
+		struct {
+			guint32 image_id;
+			guint32 status;
+		} end_of_image_transfer;
+		/* done packet = header only */
+		struct {
+			guint32 image_transfer_status;
+		} done_response;
+		/* reset packet = header only */
+		/* reset response packet = header only */
+		struct {
+			guint64 image_id;
+			guint64 offset;
+			guint64 length;
+		} read_data_64bit;
+	};
+} __attribute((packed));
+
+/* helper functions */
+static FuSaharaCommandId
+sahara_packet_get_command_id(GByteArray *packet)
+{
+	return ((struct sahara_packet *)(packet->data))->command_id;
+}
+
+static FuSaharaCommandId
+sahara_packet_get_length(GByteArray *packet)
+{
+	return ((struct sahara_packet *)(packet->data))->length;
+}
+
+/* IO functions */
+static gboolean
+fu_sahara_loader_find_interface(FuSaharaLoader *self, FuUsbDevice *usb_device, GError **error)
+{
+#if G_USB_CHECK_VERSION(0, 3, 3)
+	g_autoptr(GPtrArray) intfs = NULL;
+	GUsbDevice *g_usb_device = fu_usb_device_get_dev(usb_device);
+
+	/* all sahara devices use the same vid:pid pair */
+	if (g_usb_device_get_vid(g_usb_device) != 0x05c6 ||
+	    g_usb_device_get_pid(g_usb_device) != 0x9008) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "Wrong device and/or vendor id: 0x%04x 0x%04x",
+			    g_usb_device_get_vid(g_usb_device),
+			    g_usb_device_get_pid(g_usb_device));
+		return FALSE;
+	}
+
+	/* Parse usb interfaces and find suitable endpoints */
+	intfs = g_usb_device_get_interfaces(g_usb_device, error);
+	if (intfs == NULL)
+		return FALSE;
+	for (guint i = 0; i < intfs->len; i++) {
+		GUsbInterface *intf = g_ptr_array_index(intfs, i);
+		if (g_usb_interface_get_class(intf) == 255 &&
+		    g_usb_interface_get_subclass(intf) == 255 &&
+		    g_usb_interface_get_protocol(intf) == 255) {
```

Comment:
i think for this we've used hex before, so, `0xFF` please

Reply:
done